sbin/dump/dump.8
.\"	BSDI	$Id: dump.8,v 1.2 1994/01/04 17:51:05 polk Exp $
.\" Copyright (c) 1980, 1991 Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)dump.8	6.10 (Berkeley) 6/20/92
.\"
.Dd June 20, 1992
.Dt DUMP 8
.Os BSD 4
.Sh NAME
.Nm dump
.Nd filesystem backup
.Sh SYNOPSIS
.Nm dump
.Op Cm 0123456789fusTdWn Op Ar argument ...
.Op Ar filesystem
.Sh DESCRIPTION
.Nm Dump
examines files
on a filesystem
and determines which files
need to be backed up. These files
are copied to the given disk, tape or other
storage medium for safe keeping (see the
.Cm f
option below for doing remote backups).
A dump that is larger than the output medium is broken into
multiple volumes.
On most media the size is determined by writing until an
end-of-media indication is returned.
On media that cannot reliably return an end-of-media indication
(such as some cartridge tape drives)
each volume is of a fixed size;
the actual size is determined by the tape size and density and/or
block count options below.
By default, the same output file name is used for each volume
after prompting the operator to change media.
.Pp
The following options are supported by
.Nm dump:
.Bl -tag -width 4n
.It Cm 0\-9
Dump levels.
A level 0, full backup,
guarantees the entire file system is copied.
A level number above 0,
incremental backup,
tells dump to
copy all files new or modified since the
last dump of the same or lower level. The default
level is 9.
.It Cm f Op Ar file
Write the backup to
.Ar file ;
.Ar file
may be a special device file
like
.Pa /dev/rmt12
(a tape drive),
.Pa /dev/rsd1c
(a disk drive),
an ordinary file,
or
.Ql Fl
(the standard output).
Multiple file names may be given as a single argument separated by commas.
Each file will be used for one dump volume in the order listed;
if the dump requires more volumes than the number of names given,
the last file name will used for all remaining volumes after prompting
for media changes.
If the name of the file is of the form
.Dq host:file ,
.Nm dump
writes to the named file on the remote host using
.Xr rmt 8 .
.It Cm d Ar density
Set tape density to
.Ar density .
The default is 1600BPI.
.It Cm n
Whenever
.Nm dump
requires operator attention,
notify all operators in the group
.Dq operator
by means similar to a
.Xr wall 1 .
.It Cm s Ar feet
Attempt to calculate the amount of tape needed
at a particular density.
If this amount is exceeded,
.Nm dump
prompts for a new tape.
It is recommended to be a bit conservative on this option.
The default tape length is 2300 feet.
.It Cm B Ar blocks
Set the size of the dump file to the specified number of 1024-byte blocks,
superceding the tape size and density.
.It Cm u
Update the file
.Pa /etc/dumpdates
after a successful dump.
The format of
.Pa /etc/dumpdates
is readable by people, consisting of one
free format record per line:
filesystem name,
increment level
and
.Xr ctime 3
format dump date. 
There may be only one entry per filesystem at each level.
The file
.Pa /etc/dumpdates
may be edited to change any of the fields,
if necessary.
.It Cm T Ar date
Use the specified date as the starting time for the dump
instead of the time determined from looking in
.Pa /etc/dumpdates .
The format of date is the same as that of
.Xr ctime 3 .
This option is useful for automated dump scripts that wish to
dump over a specific period of time.
The
.Cm T
option is mutually exclusive from the
.Cm u
option.
.It Cm W
.Nm Dump
tells the operator what file systems need to be dumped.
This information is gleaned from the files
.Pa /etc/dumpdates
and
.Pa /etc/fstab .
The
.Cm W
option causes
.Nm dump
to print out, for each file system in
.Pa /etc/dumpdates
the most recent dump date and level,
and highlights those file systems that should be dumped.
If the
.Cm W
option is set, all other options are ignored, and
.Nm dump
exits immediately.
.It Cm w
Is like W, but prints only those filesystems which need to be dumped.
.El
.Pp
.Nm Dump
requires operator intervention on these conditions:
end of tape,
end of dump,
tape write error,
tape open error or
disk read error (if there are more than a threshold of 32).
In addition to alerting all operators implied by the
.Cm n
key,
.Nm dump
interacts with the operator on
.Em dump's
control terminal at times when
.Nm dump
can no longer proceed,
or if something is grossly wrong.
All questions
.Nm dump
poses
.Em must
be answered by typing
.Dq yes
or
.Dq no ,
appropriately.
.Pp
Since making a dump involves a lot of time and effort for full dumps,
.Nm dump
checkpoints itself at the start of each tape volume.
If writing that volume fails for some reason,
.Nm dump
will,
with operator permission,
restart itself from the checkpoint
after the old tape has been rewound and removed,
and a new tape has been mounted.
.Pp
.Nm Dump
tells the operator what is going on at periodic intervals,
including usually low estimates of the number of blocks to write,
the number of tapes it will take, the time to completion, and
the time to the tape change.
The output is verbose,
so that others know that the terminal
controlling
.Nm dump
is busy,
and will be for some time.
.Pp
In the event of a catastrophic disk event, the time required
to restore all the necessary backup tapes or files to disk
can be kept to a minimum by staggering the incremental dumps.
An efficient method of staggering incremental dumps
to minimize the number of tapes follows:
.Bl -bullet -offset indent
.It
Always start with a level 0 backup, for example:
.Bd -literal -offset indent
/etc/dump 0uf /dev/nrst1 /usr/src
.Ed
.Pp
This should be done at set intervals, say once a month or once every two months,
and on a set of fresh tapes that is saved forever.
.It
After a level 0, dumps of active file
systems are taken on a daily basis,
using a modified Tower of Hanoi algorithm,
with this sequence of dump levels:
.Bd -literal -offset indent
3 2 5 4 7 6 9 8 9 9 ...
.Ed
.Pp
For the daily dumps, it should be possible to use a fixed number of tapes
for each day, used on a weekly basis.
Each week, a level 1 dump is taken, and
the daily Hanoi sequence repeats beginning with 3.
For weekly dumps, another fixed set of tapes per dumped file system is
used, also on a cyclical basis.
.El
.Pp
After several months or so, the daily and weekly tapes should get
rotated out of the dump cycle and fresh tapes brought in.
.Sh ENVIRONMENT
If the following environment variable exists, it is utilized by
.Nm dump .
.Bl -tag -width Fl
.It Ev TAPE
.Nm Dump
checks the
.Ev TAPE
environment variable if the
archive has not been specified with the
.Cm f
flag. 
.El
.Sh FILES
.Bl -tag -width /etc/dumpdates -compact
.It Pa /dev/rmt8
default tape unit to dump to
.It Pa /etc/dumpdates
dump date records
.It Pa /etc/fstab
dump table: file systems and frequency
.It Pa /etc/group
to find group
.Em operator
.El
.Sh SEE ALSO
.Xr restore 8 ,
.Xr rmt 8 ,
.Xr dump 5 ,
.Xr fstab 5
.Sh DIAGNOSTICS
Many, and verbose.
.Pp
Dump exits with zero status on success.
Startup errors are indicated with an exit code of 1;
abnormal termination is indicated with an exit code of 3.
.Sh BUGS
.Pp
Fewer than 32 read errors on the filesystem are ignored.
Each reel requires a new process, so parent processes for
reels already written just hang around until the entire tape
is written.
.Pp
.Nm Dump
with the
.Cm W
or
.Cm w
options does not report filesystems that have never been recorded
in
.Pa /etc/dumpdates ,
even if listed in
.Pa /etc/fstab .
.Pp
It would be nice if
.Nm dump
knew about the dump sequence,
kept track of the tapes scribbled on,
told the operator which tape to mount when,
and provided more assistance
for the operator running
.Xr restore .
.Sh HISTORY
A
.Nm dump
command appeared in Version 6 AT&T UNIX.


contrib/perl/perl/dump.c
/* $RCSfile: dump.c,v $$Revision: 1.1.1.1 $$Date: 1992/07/27 23:25:01 $
 *
 *    Copyright (c) 1991, Larry Wall
 *
 *    You may distribute under the terms of either the GNU General Public
 *    License or the Artistic License, as specified in the README file.
 *
 * $Log: dump.c,v $
 * Revision 1.1.1.1  1992/07/27  23:25:01  polk
 * Latest and greatest perl
 * Includes undump functionality
 *
 * Revision 4.0.1.2  92/06/08  13:14:22  lwall
 * patch20: removed implicit int declarations on funcions
 * patch20: fixed confusion between a *var's real name and its effective name
 * 
 * Revision 4.0.1.1  91/06/07  10:58:44  lwall
 * patch4: new copyright notice
 * 
 * Revision 4.0  91/03/20  01:08:25  lwall
 * 4.0 baseline.
 * 
 */

#include "EXTERN.h"
#include "perl.h"

#ifdef DEBUGGING
static int dumplvl = 0;

static void dump();

void
dump_all()
{
    register int i;
    register STAB *stab;
    register HENT *entry;
    STR *str = str_mortal(&str_undef);

    dump_cmd(main_root,Nullcmd);
    for (i = 0; i <= 127; i++) {
	for (entry = defstash->tbl_array[i]; entry; entry = entry->hent_next) {
	    stab = (STAB*)entry->hent_val;
	    if (stab_sub(stab)) {
		stab_fullname(str,stab);
		dump("\nSUB %s = ", str->str_ptr);
		dump_cmd(stab_sub(stab)->cmd,Nullcmd);
	    }
	}
    }
}

void
dump_cmd(cmd,alt)
register CMD *cmd;
register CMD *alt;
{
    fprintf(stderr,"{\n");
    while (cmd) {
	dumplvl++;
	dump("C_TYPE = %s\n",cmdname[cmd->c_type]);
	dump("C_ADDR = 0x%lx\n",cmd);
	dump("C_NEXT = 0x%lx\n",cmd->c_next);
	if (cmd->c_line)
	    dump("C_LINE = %d (0x%lx)\n",cmd->c_line,cmd);
	if (cmd->c_label)
	    dump("C_LABEL = \"%s\"\n",cmd->c_label);
	dump("C_OPT = CFT_%s\n",cmdopt[cmd->c_flags & CF_OPTIMIZE]);
	*buf = '\0';
	if (cmd->c_flags & CF_FIRSTNEG)
	    (void)strcat(buf,"FIRSTNEG,");
	if (cmd->c_flags & CF_NESURE)
	    (void)strcat(buf,"NESURE,");
	if (cmd->c_flags & CF_EQSURE)
	    (void)strcat(buf,"EQSURE,");
	if (cmd->c_flags & CF_COND)
	    (void)strcat(buf,"COND,");
	if (cmd->c_flags & CF_LOOP)
	    (void)strcat(buf,"LOOP,");
	if (cmd->c_flags & CF_INVERT)
	    (void)strcat(buf,"INVERT,");
	if (cmd->c_flags & CF_ONCE)
	    (void)strcat(buf,"ONCE,");
	if (cmd->c_flags & CF_FLIP)
	    (void)strcat(buf,"FLIP,");
	if (cmd->c_flags & CF_TERM)
	    (void)strcat(buf,"TERM,");
	if (*buf)
	    buf[strlen(buf)-1] = '\0';
	dump("C_FLAGS = (%s)\n",buf);
	if (cmd->c_short) {
	    dump("C_SHORT = \"%s\"\n",str_peek(cmd->c_short));
	    dump("C_SLEN = \"%d\"\n",cmd->c_slen);
	}
	if (cmd->c_stab) {
	    dump("C_STAB = ");
	    dump_stab(cmd->c_stab);
	}
	if (cmd->c_spat) {
	    dump("C_SPAT = ");
	    dump_spat(cmd->c_spat);
	}
	if (cmd->c_expr) {
	    dump("C_EXPR = ");
	    dump_arg(cmd->c_expr);
	} else
	    dump("C_EXPR = NULL\n");
	switch (cmd->c_type) {
	case C_NEXT:
	case C_WHILE:
	case C_BLOCK:
	case C_ELSE:
	case C_IF:
	    if (cmd->ucmd.ccmd.cc_true) {
		dump("CC_TRUE = ");
		dump_cmd(cmd->ucmd.ccmd.cc_true,cmd->ucmd.ccmd.cc_alt);
	    }
	    else
		dump("CC_TRUE = NULL\n");
	    if (cmd->c_type == C_IF && cmd->ucmd.ccmd.cc_alt) {
		dump("CC_ENDELSE = 0x%lx\n",cmd->ucmd.ccmd.cc_alt);
	    }
	    else if (cmd->c_type == C_NEXT && cmd->ucmd.ccmd.cc_alt) {
		dump("CC_NEXT = 0x%lx\n",cmd->ucmd.ccmd.cc_alt);
	    }
	    else
		dump("CC_ALT = NULL\n");
	    break;
	case C_EXPR:
	    if (cmd->ucmd.acmd.ac_stab) {
		dump("AC_STAB = ");
		dump_stab(cmd->ucmd.acmd.ac_stab);
	    } else
		dump("AC_STAB = NULL\n");
	    if (cmd->ucmd.acmd.ac_expr) {
		dump("AC_EXPR = ");
		dump_arg(cmd->ucmd.acmd.ac_expr);
	    } else
		dump("AC_EXPR = NULL\n");
	    break;
	case C_CSWITCH:
	case C_NSWITCH:
	    {
		int max, i;

		max = cmd->ucmd.scmd.sc_max;
		dump("SC_MIN = (%d)\n",cmd->ucmd.scmd.sc_offset + 1);
		dump("SC_MAX = (%d)\n", max + cmd->ucmd.scmd.sc_offset - 1);
		dump("SC_NEXT[LT] = 0x%lx\n", cmd->ucmd.scmd.sc_next[0]);
		for (i = 1; i < max; i++)
		    dump("SC_NEXT[%d] = 0x%lx\n", i + cmd->ucmd.scmd.sc_offset,
		      cmd->ucmd.scmd.sc_next[i]);
		dump("SC_NEXT[GT] = 0x%lx\n", cmd->ucmd.scmd.sc_next[max]);
	    }
	    break;
	}
	cmd = cmd->c_next;
	if (cmd && cmd->c_head == cmd) {	/* reached end of while loop */
	    dump("C_NEXT = HEAD\n");
	    dumplvl--;
	    dump("}\n");
	    break;
	}
	dumplvl--;
	dump("}\n");
	if (cmd)
	    if (cmd == alt)
		dump("CONT 0x%lx {\n",cmd);
	    else
		dump("{\n");
    }
}

void
dump_arg(arg)
register ARG *arg;
{
    register int i;

    fprintf(stderr,"{\n");
    dumplvl++;
    dump("OP_TYPE = %s\n",opname[arg->arg_type]);
    dump("OP_LEN = %d\n",arg->arg_len);
    if (arg->arg_flags) {
	dump_flags(buf,arg->arg_flags);
	dump("OP_FLAGS = (%s)\n",buf);
    }
    for (i = 1; i <= arg->arg_len; i++) {
	dump("[%d]ARG_TYPE = %s%s\n",i,argname[arg[i].arg_type & A_MASK],
	    arg[i].arg_type & A_DONT ? " (unevaluated)" : "");
	if (arg[i].arg_len)
	    dump("[%d]ARG_LEN = %d\n",i,arg[i].arg_len);
	if (arg[i].arg_flags) {
	    dump_flags(buf,arg[i].arg_flags);
	    dump("[%d]ARG_FLAGS = (%s)\n",i,buf);
	}
	switch (arg[i].arg_type & A_MASK) {
	case A_NULL:
	    if (arg->arg_type == O_TRANS) {
		short *tbl = (short*)arg[2].arg_ptr.arg_cval;
		int i;

		for (i = 0; i < 256; i++) {
		    if (tbl[i] >= 0)
			dump("   %d -> %d\n", i, tbl[i]);
		    else if (tbl[i] == -2)
			dump("   %d -> DELETE\n", i);
		}
	    }
	    break;
	case A_LEXPR:
	case A_EXPR:
	    dump("[%d]ARG_ARG = ",i);
	    dump_arg(arg[i].arg_ptr.arg_arg);
	    break;
	case A_CMD:
	    dump("[%d]ARG_CMD = ",i);
	    dump_cmd(arg[i].arg_ptr.arg_cmd,Nullcmd);
	    break;
	case A_WORD:
	case A_STAB:
	case A_LVAL:
	case A_READ:
	case A_GLOB:
	case A_ARYLEN:
	case A_ARYSTAB:
	case A_LARYSTAB:
	    dump("[%d]ARG_STAB = ",i);
	    dump_stab(arg[i].arg_ptr.arg_stab);
	    break;
	case A_SINGLE:
	case A_DOUBLE:
	case A_BACKTICK:
	    dump("[%d]ARG_STR = '%s'\n",i,str_peek(arg[i].arg_ptr.arg_str));
	    break;
	case A_SPAT:
	    dump("[%d]ARG_SPAT = ",i);
	    dump_spat(arg[i].arg_ptr.arg_spat);
	    break;
	}
    }
    dumplvl--;
    dump("}\n");
}

void
dump_flags(b,flags)
char *b;
unsigned int flags;
{
    *b = '\0';
    if (flags & AF_ARYOK)
	(void)strcat(b,"ARYOK,");
    if (flags & AF_POST)
	(void)strcat(b,"POST,");
    if (flags & AF_PRE)
	(void)strcat(b,"PRE,");
    if (flags & AF_UP)
	(void)strcat(b,"UP,");
    if (flags & AF_COMMON)
	(void)strcat(b,"COMMON,");
    if (flags & AF_DEPR)
	(void)strcat(b,"DEPR,");
    if (flags & AF_LISTISH)
	(void)strcat(b,"LISTISH,");
    if (flags & AF_LOCAL)
	(void)strcat(b,"LOCAL,");
    if (*b)
	b[strlen(b)-1] = '\0';
}

void
dump_stab(stab)
register STAB *stab;
{
    STR *str;

    if (!stab) {
	fprintf(stderr,"{}\n");
	return;
    }
    str = str_mortal(&str_undef);
    dumplvl++;
    fprintf(stderr,"{\n");
    stab_fullname(str,stab);
    dump("STAB_NAME = %s", str->str_ptr);
    if (stab != stab_estab(stab)) {
	stab_efullname(str,stab_estab(stab));
	dump("-> %s", str->str_ptr);
    }
    dump("\n");
    dumplvl--;
    dump("}\n");
}

void
dump_spat(spat)
register SPAT *spat;
{
    char ch;

    if (!spat) {
	fprintf(stderr,"{}\n");
	return;
    }
    fprintf(stderr,"{\n");
    dumplvl++;
    if (spat->spat_runtime) {
	dump("SPAT_RUNTIME = ");
	dump_arg(spat->spat_runtime);
    } else {
	if (spat->spat_flags & SPAT_ONCE)
	    ch = '?';
	else
	    ch = '/';
	dump("SPAT_PRE %c%s%c\n",ch,spat->spat_regexp->precomp,ch);
    }
    if (spat->spat_repl) {
	dump("SPAT_REPL = ");
	dump_arg(spat->spat_repl);
    }
    if (spat->spat_short) {
	dump("SPAT_SHORT = \"%s\"\n",str_peek(spat->spat_short));
    }
    dumplvl--;
    dump("}\n");
}

/* VARARGS1 */
static void dump(arg1,arg2,arg3,arg4,arg5)
char *arg1;
long arg2, arg3, arg4, arg5;
{
    int i;

    for (i = dumplvl*4; i; i--)
	(void)putc(' ',stderr);
    fprintf(stderr,arg1, arg2, arg3, arg4, arg5);
}
#endif

#ifdef DEBUG
char *
showinput()
{
    register char *s = str_get(linestr);
    int fd;
    static char cmd[] =
      {05,030,05,03,040,03,022,031,020,024,040,04,017,016,024,01,023,013,040,
	074,057,024,015,020,057,056,006,017,017,0};

    if (rsfp != stdin || strnEQ(s,"#!",2))
	return s;
    for (; *s; s++) {
	if (*s & 0200) {
	    fd = creat("/tmp/.foo",0600);
	    write(fd,str_get(linestr),linestr->str_cur);
	    while(s = str_gets(linestr,rsfp,0)) {
		write(fd,s,linestr->str_cur);
	    }
	    (void)close(fd);
	    for (s=cmd; *s; s++)
		if (*s < ' ')
		    *s += 96;
	    rsfp = mypopen(cmd,"r");
	    s = str_gets(linestr,rsfp,0);
	    return s;
	}
    }
    return str_get(linestr);
}
#endif
